### ISSUE TITLE
[Docs]: Improve documentation for possible table structures

### ISSUE THREAD

**Issue by mrjones2014** [mrjones2014/legendary.nvim#106]
### Similar Issues

- [X] Before filing, I have searched for similar issues.

### Description

There's a few different table structures that are accepted now. In particular, keymaps accept several different structures. They're all documented, but not super well. I see two main problems.

1. The documentation that is there is too wordy. We need cold hard API documentation, with types. Honestly I'm considering rewriting parts in Teal so I can add static types and use a documentation generator for this.

2. There needs to be a brief overview of the power of what you can do with `legendary.nvim` closer to the top of the README.md

@tsyroid @olimorris thoughts and input always appreciated.

**Comment by mrjones2014** [comment]
Probably also separate docs for table structures vs. helper functions, currently they're kinda both covered in the same section.

**Comment by tsyroid** [comment]
Mat (et al),

Thanks for including me in this dialogue.

I’ll respond to this request for comment, but it’ll take me a day or two.

I’m currently neck-deep in converting all my existing keymaps to legendary
format. Wish I had caught this wave earlier, but it was clearly not to be.
But hopefully I’ll be  able to offer some transitory feedback in the
meantime…

Once I’m “there” (functional), I’ll go back over the existing documentation
and help where I can.

Best,
/tom

**Comment by mrjones2014** [comment]
Teal type definitions for the `vim` global: https://github.com/teal-language/teal-types/tree/master/types/neovim

**Comment by mrjones2014** [comment]
Now that we've got Teal types, we should look into a way to auto-generate some stuff from the Teal types. The Teal types are more accurate than the existing annotations were.

**Comment by tsyroid** [comment]
Define “some stuff” :smirk:

No, let’s back up a peg. What kinda functionality do you want to add to
Legendary that is not yet present?

(With the above in hand, “some stuff” ~~will~~ should magically unveil
itself to me)

**Comment by mrjones2014** [comment]
What I want ideally is for the table structures to be "self-documenting", that is, when the Teal types change, it auto-generates new documentation on the table structures for each type. Basically the whole "Table Structures" section of the README.md can/should be "self-documenting" from the Teal types.

I'm not sure what kind of type information compiled Teal gives at runtime, if it gives the `record` definition we could use Lua to generate the docs, if not we'll have to build in some functionality to Teal somehow or parse the Teal source.

**Comment by mrjones2014** [comment]
TJ has a project with a Lua docgen module 👀 
https://github.com/tjdevries/tree-sitter-lua/tree/master/lua/docgen

**Comment by mrjones2014** [comment]
Another tool that could help

https://github.com/numToStr/lemmy-help

**Comment by tsyroid** [comment]
I’d talk to the author.

If you look at his mini.nvim help file (which I’ve spent a fair amount of
time in), you’ll see he has headers.

https://github.com/echasnovski/mini.nvim/blob/main/doc/mini.txt

**Comment by tsyroid** [comment]
You can install it that way I suppose. Not the way he intended it (I don’t
think). I have in the past simply plucked out the module I needed, dropped
it into lua/modules and called it from init.lua.

Not the point, though. Sorry, I got my turds wisted because I happed to
have the mini.doc open at the time.

Ask the author of lemmy-help (num2Str) if there's an easy way to parse in
headers. I see he doesn’t eat his own dog food though (at least not at this
time); most of his help files for the apps he’s written are  markdown.

**Comment by mrjones2014** [comment]
Yup we're chatting about it in this issue https://github.com/numToStr/lemmy-help/issues/1

**Comment by mrjones2014** [comment]
That issue's had a PR merged! Going to start on this soon.

**Comment by mrjones2014** [comment]
Got decent API documentation generation in #140 but I need to figure out how/where to insert them into `doc/legendary.txt`.

**Comment by mrjones2014** [comment]
You can view how the docs will be generated by running `make api-docs`. Currently it just prints the docs to the terminal.

**Comment by tsyroid** [comment]
I guess my shortly wasn’t shortly enough.

Just pulled and sync’d. Not there now. My HEAD/master is at 77b7bc3.

Any idea what commit I need to be on to see this? I’d also like to see your
Makefile code for this action.

**Comment by mrjones2014** [comment]
You can grab the branch and look at the diff from the PR: https://github.com/mrjones2014/legendary.nvim/pull/140

`git checkout matjones/106-improved-docs`

**Comment by tsyroid** [comment]
Sorry. Found it. Didn’t know you had a branch happening.

**Comment by tsyroid** [comment]
Close ;-)
(reminds me of a joke about hand grenades…)

git pull upstream matjones/106-improved-docs

I like it. I particularly like how you set things up to test. Makefile ->
output to CLI to see what’s happening.

**Comment by mrjones2014** [comment]
Note to self: document the issue linked above ☝️
